### ISSUE TITLE
[BUG][v1.7.0-rc1] Workload unable to recover after kubelet restart with error: MountVolume.MountDevice failed

### ISSUE THREAD

**Issue by yangchiu** [longhorn/longhorn#9014]
## Describe the bug

Negative test case `Restart Volume Node Kubelet While Workload Heavy Writing` failed on `v1.7.0-rc1` and `master-head`.

After stopped the `k3s-agent` of the volume attached node for 10 seconds and restarted it, the workload with a RWX volume could be unable to recover and get stuck in `ContainerCreating` state forever:

```
# kubectl get pods -owide
NAME                                   READY   STATUS              RESTARTS   AGE     IP            NODE            NOMINATED NODE   READINESS GATES
longhorn-test                          2/2     Running             0          10h     10.42.0.2     ip-10-0-1-120   <none>           <none>
longhorn-test-minio-7d8ddc774f-gmzwt   1/1     Running             0          7h21m   10.42.1.169   ip-10-0-2-125   <none>           <none>
longhorn-test-nfs-5f96f8f5f8-b2nk4     1/1     Running             0          7h21m   10.42.1.186   ip-10-0-2-125   <none>           <none>
e2e-test-statefulset-1-0               0/1     ContainerCreating   0          7h15m   <none>        ip-10-0-2-92    <none>           <none>
e2e-test-statefulset-0-0               1/1     Running             0          7h9m    10.42.2.66    ip-10-0-2-187   <none>           <none>
```

With `FailedMount` error:

```
# kubectl describe pod e2e-test-statefu
lset-1-0
Name:             e2e-test-statefulset-1-0
Namespace:        default
Priority:         0
Service Account:  default
Node:             ip-10-0-2-92/10.0.2.92
Start Time:       Tue, 16 Jul 2024 17:32:04 +0000
Labels:           app=e2e-test-statefulset-1
                  apps.kubernetes.io/pod-index=0
                  controller-revision-hash=e2e-test-statefulset-1-7d6d94679d
                  statefulset.kubernetes.io/pod-name=e2e-test-statefulset-1-0
Annotations:      <none>
Status:           Pending
IP:               
IPs:              <none>
Controlled By:    StatefulSet/e2e-test-statefulset-1
Containers:
  sleep:
    Container ID:  
    Image:         busybox:1.34.0
    Image ID:      
    Port:          <none>
    Host Port:     <none>
    Args:
      /bin/sh
      -c
      while true;do date;sleep 5; done
    State:          Waiting
      Reason:       ContainerCreating
    Ready:          False
    Restart Count:  0
    Environment:    <none>
    Mounts:
      /data from pod-data (rw)
      /var/run/secrets/kubernetes.io/serviceaccount from kube-api-access-9dsqw (ro)
Conditions:
  Type                        Status
  PodReadyToStartContainers   False 
  Initialized                 True 
  Ready                       False 
  ContainersReady             False 
  PodScheduled                True 
Volumes:
  pod-data:
    Type:       PersistentVolumeClaim (a reference to a PersistentVolumeClaim in the same namespace)
    ClaimName:  pod-data-e2e-test-statefulset-1-0
    ReadOnly:   false
  kube-api-access-9dsqw:
    Type:                    Projected (a volume that contains injected data from multiple sources)
    TokenExpirationSeconds:  3607
    ConfigMapName:           kube-root-ca.crt
    ConfigMapOptional:       <nil>
    DownwardAPI:             true
QoS Class:                   BestEffort
Node-Selectors:              <none>
Tolerations:                 node.kubernetes.io/not-ready:NoExecute op=Exists for 300s
                             node.kubernetes.io/unreachable:NoExecute op=Exists for 300s
Events:
  Type     Reason       Age                     From     Message
  ----     ------       ----                    ----     -------
  Warning  FailedMount  5m12s (x54 over 6h57m)  kubelet  MountVolume.MountDevice failed for volume "pvc-613b3fdc-6f28-43cd-93f9-c92257177404" : kubernetes.io/csi: attacher.MountDevice failed to determine if the node service has VOLUME_MOUNT_GROUP capability: rpc error: code = DeadlineExceeded desc = context deadline exceeded
```

The nodes and volumes seem to be working well:

```
# kubectl get nodes -owide
NAME            STATUS   ROLES                  AGE   VERSION        INTERNAL-IP   EXTERNAL-IP   OS-IMAGE                              KERNEL-VERSION              CONTAINER-RUNTIME
ip-10-0-1-120   Ready    control-plane,master   10h   v1.30.0+k3s1   10.0.1.120    <none>        SUSE Linux Enterprise Server 15 SP6   6.4.0-150600.23.7-default   containerd://1.7.15-k3s1
ip-10-0-2-125   Ready    <none>                 10h   v1.30.0+k3s1   10.0.2.125    <none>        SUSE Linux Enterprise Server 15 SP6   6.4.0-150600.23.7-default   containerd://1.7.15-k3s1
ip-10-0-2-92    Ready    <none>                 10h   v1.30.0+k3s1   10.0.2.92     <none>        SUSE Linux Enterprise Server 15 SP6   6.4.0-150600.23.7-default   containerd://1.7.15-k3s1
ip-10-0-2-187   Ready    <none>                 10h   v1.30.0+k3s1   10.0.2.187    <none>        SUSE Linux Enterprise Server 15 SP6   6.4.0-150600.23.7-default   containerd://1.7.15-k3s1
```

```
# kubectl get volumes -n longhorn-system -owide
NAME                                       DATA ENGINE   STATE      ROBUSTNESS   SCHEDULED   SIZE         NODE            AGE
pvc-613b3fdc-6f28-43cd-93f9-c92257177404   v1            attached   healthy                  3221225472   ip-10-0-2-125   9h
pvc-6b88e36b-dc60-4eb4-a30d-c816365436fb   v1            attached   healthy                  3221225472   ip-10-0-2-187   9h
```

The test case is like:
```
    Given Create statefulset 0 using RWO volume
    And Create statefulset 1 using RWX volume

    FOR    ${i}    IN RANGE    ${LOOP_COUNT}
        And Keep writing data to pod of statefulset 0
        And Keep writing data to pod of statefulset 1

        When Stop volume nodes kubelet for 10 seconds    statefulset 0    statefulset 1
        And Wait for volume of statefulset 0 healthy
        And Wait for volume of statefulset 1 healthy
        And Wait for workloads pods stable    statefulset 0    statefulset 1

        Then Check statefulset 0 works
        And Check statefulset 1 works
    END
```

## To Reproduce

Run negative test case `Restart Volume Node Kubelet While Workload Heavy Writing` repeatedly like:
```
-s \"kubelet_restart\" -v LOOP_COUNT:20 -v RETRY_COUNT:259200
``` 

## Support bundle for troubleshooting

[supportbundle_4446a216-b312-4df0-a075-667dbef83f5f_2024-07-17T00-47-56Z.zip](https://github.com/user-attachments/files/16257935/supportbundle_4446a216-b312-4df0-a075-667dbef83f5f_2024-07-17T00-47-56Z.zip)

## Environment

 - Longhorn version: v1.7.0-rc1 or master-head
 - Impacted volume (PV): 
 - Installation method (e.g. Rancher Catalog App/Helm/Kubectl): helm
 - Kubernetes distro (e.g. RKE/K3s/EKS/OpenShift) and version: v1.30.0+k3s1
   - Number of control plane nodes in the cluster:
   - Number of worker nodes in the cluster:
 - Node config
   - OS type and version: sles 15-sp6
   - Kernel version:
   - CPU per node:
   - Memory per node:
   - Disk type (e.g. SSD/NVMe/HDD):
   - Network bandwidth between the nodes (Gbps):
 - Underlying Infrastructure (e.g. on AWS/GCE, EKS/GKE, VMWare/KVM, Baremetal):
 - Number of Longhorn volumes in the cluster:

**Comment by c3y1huang** [comment]
Analyse (WIP):

> MountVolume.MountDevice failed for volume "pvc-613b3fdc-6f28-43cd-93f9-c92257177404" : kubernetes.io/csi: attacher.MountDevice failed to determine if the node service has VOLUME_MOUNT_GROUP capability: rpc error: code = DeadlineExceeded desc = context deadline exceeded

https://github.com/kubernetes/kubernetes/blob/v1.30.0/pkg/volume/csi/csi_attacher.go#L394

**Comment by c3y1huang** [comment]
@yangchiu , I have some questions need your help to clarify:
1. Is this reproducible when running only `Restart Volume Node Kubelet While Workload Heavy Writing`?
    > -s \"kubelet_restart\" -v LOOP_COUNT:20 -v RETRY_COUNT:259200
1. Is this reproducible on v1.6.x with the same test image?

**Comment by yangchiu** [comment]
> @yangchiu , I have some questions that need your help to clarify:
> 
>     1. ~There seem to be 2 tests in the suit. Is this reproducible when running only `Restart Volume Node Kubelet While Workload Heavy Writing`?~ Answering my own question: From the log, it seems to be hanging when running `Stop Volume Node Kubelet For More Than Pod Eviction Timeout While Workload Heavy Writing`. So, running `Restart Volume Node Kubelet While Workload Heavy Writing` is unlikely to reproduce the issue.

Both https://ci.longhorn.io/job/private/job/longhorn-e2e-test/757/ and https://ci.longhorn.io/job/private/job/longhorn-e2e-test/772/encountered this issue during `Restart Volume Node Kubelet While Workload Heavy Writing`.

> 
>     2. Is this reproducible on v1.6.x with the same test image?
>        WIP, https://ci.longhorn.io/job/private/job/longhorn-e2e-test/773/console

Can't reproduce so far.

> 
>     3. We made a fix in [[BUG][v1.7.0-rc1] Require unexpected long time to remount RWX volume after share manager restarted #8999](https://github.com/longhorn/longhorn/issues/8999) and I am not sure if it's related. Could you help to run the test again to see if this issue is reproducible?
>        Yes, https://ci.longhorn.io/job/private/job/longhorn-e2e-test/772/

Yes, it's still reproducible on `v1.7.x-head` in https://ci.longhorn.io/job/private/job/longhorn-e2e-test/772/:

```
/src/longhorn-tests/test_framework/terraform/aws/sles # kubectl get nodes
NAME            STATUS   ROLES                  AGE   VERSION
ip-10-0-2-254   Ready    <none>                 19h   v1.30.0+k3s1
ip-10-0-1-42    Ready    control-plane,master   19h   v1.30.0+k3s1
ip-10-0-2-131   Ready    <none>                 19h   v1.30.0+k3s1
ip-10-0-2-101   Ready    <none>                 19h   v1.30.0+k3s1
/src/longhorn-tests/test_framework/terraform/aws/sles # kubectl get pods 
NAME                                   READY   STATUS              RESTARTS   AGE
longhorn-test                          2/2     Running             0          19h
longhorn-test-nfs-5f96f8f5f8-tqhjw     1/1     Running             0          13h
longhorn-test-minio-7d8ddc774f-wtw4q   1/1     Running             0          13h
e2e-test-statefulset-1-0               0/1     ContainerCreating   0          13h
e2e-test-statefulset-0-0               1/1     Running             0          13h
/src/longhorn-tests/test_framework/terraform/aws/sles # kubectl get volumes -n longhorn-system
NAME                                       DATA ENGINE   STATE      ROBUSTNESS   SCHEDULED   SIZE         NODE            AGE
pvc-08bacca7-cdd1-4043-a65c-6f2f1e3cb917   v1            attached   healthy                  3221225472   ip-10-0-2-254   17h
pvc-d84e55b6-4552-4413-95ff-9abf46424919   v1            attached   healthy                  3221225472   ip-10-0-2-101   17h
/src/longhorn-tests/test_framework/terraform/aws/sles # kubectl describe pod e2e-test-statefulset-1
-0
Name:             e2e-test-statefulset-1-0
Namespace:        default
Priority:         0
Service Account:  default
Node:             ip-10-0-2-131/10.0.2.131
Start Time:       Wed, 17 Jul 2024 11:17:17 +0000
Labels:           app=e2e-test-statefulset-1
                  apps.kubernetes.io/pod-index=0
                  controller-revision-hash=e2e-test-statefulset-1-7d6d94679d
                  statefulset.kubernetes.io/pod-name=e2e-test-statefulset-1-0
Annotations:      <none>
Status:           Pending
IP:               
IPs:              <none>
Controlled By:    StatefulSet/e2e-test-statefulset-1
Containers:
  sleep:
    Container ID:  
    Image:         busybox:1.34.0
    Image ID:      
    Port:          <none>
    Host Port:     <none>
    Args:
      /bin/sh
      -c
      while true;do date;sleep 5; done
    State:          Waiting
      Reason:       ContainerCreating
    Ready:          False
    Restart Count:  0
    Environment:    <none>
    Mounts:
      /data from pod-data (rw)
      /var/run/secrets/kubernetes.io/serviceaccount from kube-api-access-2sf96 (ro)
Conditions:
  Type                        Status
  PodReadyToStartContainers   False 
  Initialized                 True 
  Ready                       False 
  ContainersReady             False 
  PodScheduled                True 
Volumes:
  pod-data:
    Type:       PersistentVolumeClaim (a reference to a PersistentVolumeClaim in the same namespace)
    ClaimName:  pod-data-e2e-test-statefulset-1-0
    ReadOnly:   false
  kube-api-access-2sf96:
    Type:                    Projected (a volume that contains injected data from multiple sources)
    TokenExpirationSeconds:  3607
    ConfigMapName:           kube-root-ca.crt
    ConfigMapOptional:       <nil>
    DownwardAPI:             true
QoS Class:                   BestEffort
Node-Selectors:              <none>
Tolerations:                 node.kubernetes.io/not-ready:NoExecute op=Exists for 300s
                             node.kubernetes.io/unreachable:NoExecute op=Exists for 300s
Events:
  Type     Reason       Age                  From     Message
  ----     ------       ----                 ----     -------
  Warning  FailedMount  17s (x102 over 13h)  kubelet  MountVolume.MountDevice failed for volume "pvc-d84e55b6-4552-4413-95ff-9abf46424919" : kubernetes.io/csi: attacher.MountDevice failed to determine if the node service has VOLUME_MOUNT_GROUP capability: rpc error: code = DeadlineExceeded desc = context deadline exceeded
```

[supportbundle_b6e53c0d-710d-4479-8c06-8df99fc94335_2024-07-18T00-54-21Z.zip](https://github.com/user-attachments/files/16276287/supportbundle_b6e53c0d-710d-4479-8c06-8df99fc94335_2024-07-18T00-54-21Z.zip)

**Comment by yangchiu** [comment]
Verification passed on master-head (longhorn-manager 2697ea5) and v1.7.x-head (longhorn-manager 3bc35a5). Test case `Restart Volume Node Kubelet While Workload Heavy Writing` passed on corresponding Jenkins pipelines.
